### PR TITLE
Notification on deployment finish

### PR DESCRIPTION
**Pull request by mproch**

### CODE REVIEW

File: engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala [hunk 1]
```diff
@@ -56,22 +56,35 @@ class DevelopmentDeploymentManager(actorSystem: ActorSystem) extends DeploymentM
...
     logger.debug(s"Starting deploying scenario: ${processVersion.processName}..")
+    val previous = memory.get(processVersion.processName)
     val duringDeployStateStatus = createAndSaveProcessState(DuringDeploy, processVersion)
     val result = Promise[Option[ExternalDeploymentId]]()
     actorSystem.scheduler.scheduleOnce(sleepingTimeSeconds, new Runnable {
       override def run(): Unit = {
         logger.debug(s"Finished deploying scenario: ${processVersion.processName}.")
-        result.complete(Success(duringDeployStateStatus.deploymentId))
-        asyncChangeState(processVersion.processName, Running)
+        if (description(canonicalProcess).contains("deployFail")) {
```
Comment (dswiecki) [inline engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala:78]: from where comes this "deployFail" string ?

File: ui/server/src/main/resources/defaultUiConfig.conf [hunk 2]
```diff
@@ -193,3 +193,7 @@ testDataSettings: {
   testDataMaxBytes: 200000
   resultsMaxBytes: 50000000
 }
+
+notifications {
+  duration: 24 hours
```
Comment (dswiecki) [inline ui/server/src/main/resources/defaultUiConfig.conf:198]: I wonder why its exactly 24 hours ?

File: ui/server/src/main/scala/pl/touk/nussknacker/ui/notifications/NotificationService.scala [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+      case Success(None) => logger.error(s"Failed to retrieve scenario name for id: ${event.processId}")
+      case Success(Some(scenarioName)) => synchronized {
+        data.append(NotificationEvent(UUID.randomUUID().toString, event, now, user, scenarioName))
+      }
+    }
+    filterOldNotifications(now)
+  }
+
+  private def filterOldNotifications(now: Instant): Unit = synchronized {
+    data.zipWithIndex.filter(_._1.date.isBefore(now.minus(config.duration.toMillis, ChronoUnit.MILLIS))).foreach(i => data.remove(i._2))
```
Comment (dswiecki) [inline ui/server/src/main/scala/pl/touk/nussknacker/ui/notifications/NotificationService.scala:44]: will it work on cloud with different sever and client timezones ?

File: ui/server/src/main/scala/pl/touk/nussknacker/ui/notifications/NotificationService.scala [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        .map { case (k, v) => currentDeploymentToNotification(k, v) }.toList
+    }
+  }
+
+  private def currentDeploymentToNotification(processName: ProcessName, deploymentInfo: DeployInfo): Notification = {
+    val actionString = deploymentInfo.action match {
+      case DeploymentActionType.Deployment => "deployed"
+      case DeploymentActionType.Cancel => "cancelled"
+    }
+    //TODO: should it be displayed only once?
```
Comment (dswiecki) [inline ui/server/src/main/scala/pl/touk/nussknacker/ui/notifications/NotificationService.scala:97]: meaning right now it's shown multiple times ?

File: ui/server/src/main/scala/pl/touk/nussknacker/ui/notifications/NotificationService.scala (hunk 3 above)
Comment (mproch) [reply ui/server/src/main/scala/pl/touk/nussknacker/ui/notifications/NotificationService.scala:44]: we use Instant here, no time zones.

File: ui/server/src/main/scala/pl/touk/nussknacker/ui/notifications/NotificationService.scala (hunk 4 above)
Comment (mproch) [reply ui/server/src/main/scala/pl/touk/nussknacker/ui/notifications/NotificationService.scala:97]: it keeps popping up, until deployment finishes - I've kept current behaviour

File: ui/server/src/main/resources/defaultUiConfig.conf (hunk 2 above)
Comment (mproch) [reply ui/server/src/main/resources/defaultUiConfig.conf:198]: "quite a long time, but not forever" :) do you have better value for default?

File: engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala (hunk 1 above)
Comment (mproch) [reply engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala:78]: from here - I can extract it to constant

File: ui/client/reducers/selectors/other.ts [hunk 5]
```diff
@@ -18,3 +21,6 @@ export const getCapabilities = createSelector(
   }),
 )
 
+export const getNotifications = createSelector((state: RootState): NotificationsState => state.notifications, not => not)
```
Comment (JulianWielga) [inline ui/client/reducers/selectors/other.ts:24]: reselect is unnecessary here

File: ui/client/actions/nk/notifications.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+export type NotificationActions =
+  | UpdateBackendNotifications
+  | MarkBackendNotificationRead
+
+export type UpdateBackendNotifications = {
+    type: "UPDATE_BACKEND_NOTIFICATIONS",
+    notificationIds: string[]
+}
+
+export function updateBackendNotifications(notificationIds: string[]): ThunkAction {
```
Comment (JulianWielga) [inline ui/client/actions/nk/notifications.ts:12]: no side effects so simple action creator without dispatch is enough here

**Review by JulianWielga — APPROVED** [review]

File: engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala (hunk 1 above)
Comment (dswiecki) [reply engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala]: Ok now I get how it works

File: ui/client/actions/nk/notifications.ts (hunk 6 above)
Comment (mproch) [reply ui/client/actions/nk/notifications.ts]: +1 true...

File: ui/server/src/main/scala/pl/touk/nussknacker/ui/notifications/NotificationService.scala [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    filterOldNotifications(now)
+  }
+
+  private def filterOldNotifications(now: Instant): Unit = synchronized {
+    data.zipWithIndex.filter(_._1.date.isBefore(now.minus(config.duration.toMillis, ChronoUnit.MILLIS))).foreach(i => data.remove(i._2))
+  }
+
+  private[notifications] def dataFor(user: LoggedUser, notificationsAfter: Option[Instant]): List[NotificationEvent] = {
+    filterOldNotifications(Instant.now(clock))
+    data.filter(event => event.user.id == user.id && !notificationsAfter.exists(_.isAfter(event.date))).toList
```
Comment (gadomsky) [inline ui/server/src/main/scala/pl/touk/nussknacker/ui/notifications/NotificationService.scala:49]: shouldn't we also synchronize "read" access?

Comment (mproch) [reply ui/server/src/main/scala/pl/touk/nussknacker/ui/notifications/NotificationService.scala]: :+1: maybe it'd be safer...

**Review by dswiecki — APPROVED** [review]

**Pull request merged**